domain/resource/service/resource.go

```diff
@@ -548,6 +563,71 @@ func (s *Service) SetRepositoryResources(
 	return s.st.SetRepositoryResources(ctx, args)
 }
 
+// AddResourcesBeforeApplication adds the details of which resource
+// revision to use before the application exists in the model. The
+// charm and resource metadata must exist. These resources are resolved
+// when the application is created using the returned Resource UUIDs.
+//
+// The following error types can be expected to be returned:
+//   - [coreerrors.NotValid] is returned if the Charm ID is not valid.
+//   - [resourceerrors.ArgumentNotValid] is returned if the origin is store and the revision
+//     is empty.
+//   - [resourceerrors.ResourceNameNotValid] is returned if resource name is empty.
+//   - [resourceerrors.ApplicationNameNotFound] if the specified application does
+//     not exist.
+func (s *Service) AddResourcesBeforeApplication(ctx context.Context, arg resource.AddResourcesBeforeApplicationArgs) ([]coreresource.UUID, error) {
+	if err := arg.CharmUUID.Validate(); err != nil {
+		return nil, errors.Errorf("charm id: %w", err)
+	}
+	if !isValidApplicationName(arg.ApplicationName) {
+		return nil, errors.Errorf("application name : %w", resourceerrors.ApplicationNameNotValid)
+	}
+	for _, res := range arg.ResourceDetails {
+		if res.Name == "" {
+			return nil, errors.Errorf("resource name is empty: %w", resourceerrors.ResourceNameNotValid)
+		}
+		if res.Origin == charmresource.OriginStore && res.Revision == nil {
+			return nil, errors.Errorf("revision is empty for store resource: %w", resourceerrors.ArgumentNotValid)
+		}
+		if res.Origin == charmresource.OriginUpload && res.Revision != nil {
+			return nil, errors.Errorf("revision is set for upload resource: %w", resourceerrors.ArgumentNotValid)
+		}
+	}
+	resourceUUIDs, err := s.st.AddResourcesBeforeApplication(ctx, arg)
+	if err != nil {
+		return nil, errors.Errorf("failed to add resources: %w", err)
+	}
+	return resourceUUIDs, nil
+}
+
+// UpdateStoreResource updates the revision of a store resource to a new
+// version. Increments charm modified version for the application to
+// trigger use of the new resource revision by the application.
+//
+// The following error types can be expected to be returned:
+//   - [coreerrors.NotValid] is returned if the Resource ID is not valid.
+func (s *Service) UpdateStoreResource(ctx context.Context, arg resource.UpdateStoreResourceArgs) error {
```

Comment:
I think this method should work similarly to `StoreResource` in that we should also be changing the blob here since this is also a service level operation and needs to version.

On the other hand, is this method actually needed at all? StoreResource sets the revision, and it removes old blobs and resource records. What is different about this method to that?

Reply:
The blob does not change in this case. Thus a different service method.

This method will trigger an update for a resource to be retrieved from CharmHub, not one which can been uploaded.

Reply:
Ah I see, for the name it could be `UpdateCharmhubResource`, `UpdateOriginStoreResource`, or even just `UpdateResourceRevision`.

This method should also unlink and remove the old resource blob if it is updating the revision, otherwise we have the incorrect storage information attached to the resource. This will also mean that if the uniter tries to open the resource the resource Opener will download a resource from Charmhub rather than opening the existing resource.

Reply:
Let's avoid using Charmhub in domain method names. It's the current repo and may change in the future.